Publish event for app token create/update/delete

kesselb opened the pull request:
Close https://github.com/nextcloud/server/issues/10964

![image](https://user-images.githubusercontent.com/3902676/51428572-6f589000-1c05-11e9-916b-b0ce792afa36.png)

Trigger event for activity app if a user create, update or delete an app token. ~~Unfortunately there is no easy way to inject the app token name.~~

Possible enhancement: Add tokenName as valid definition https://github.com/nextcloud/server/blob/master/lib/public/RichObjectStrings/Definitions.php for subject. (@nickvergessen is there an easier way?)

kesselb committed:
Publish event for app token create/update/delete

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

danielkesselberg commented:
https://github.com/nextcloud/server/pull/13548/commits/8aab4601b2bc2b4441500ee9d240dc6ccc4e93b4 no typehint vs `?string` what do you prefer? With `string` these test fails https://github.com/nextcloud/server/blob/720ef48d9283b53356cfadfb56adda13bfa96e64/tests/Settings/ApplicationTest.php#L72

ChristophWurst commented on settings/Controller/AuthSettingsController.php:
```diff
@@ -204,6 +221,25 @@ public function update($id, array $scope) {
+			$this->activityManager->publish($event);
+		} catch (BadMethodCallException $e) {
+			$this->logger->warning('could not publish activity', ['app' => 'settings']);
+			$this->logger->logException($e, ['app' => 'settings']);
```
I think the app context might not be necessary. @rullzer do we inject an app logger for settings as well?

rullzer replied:
try it ;) I think so

ChristophWurst reviewed (changes requested):
Code looks good! I like this enhancement!

ChristophWurst commented:
> [8aab460](https://github.com/nextcloud/server/commit/8aab4601b2bc2b4441500ee9d240dc6ccc4e93b4) no typehint vs `?string` what do you prefer? With `string` these test fails

I like `?string` :)

nickvergessen commented on settings/Controller/AuthSettingsController.php:
```diff
@@ -204,6 +221,25 @@ public function update($id, array $scope) {
+	private function publishActivity(string $subject): void {
+		$event = $this->activityManager->generateEvent();
+		$event->setApp('settings')
+			->setType('security')
```
mind also setting the object data?
`->setObject('app_token', $id)`

In the future we can then jump to activity done on the same object.

nickvergessen commented on settings/Controller/AuthSettingsController.php:
```diff
@@ -204,6 +221,25 @@ public function update($id, array $scope) {
+			->setType('security')
+			->setAffectedUser($this->uid)
+			->setAuthor($this->uid)
+			->setSubject($subject);
```
Would it make sense to include the name into the message, so you can see which token was created and maybe delete it more easily?

nickvergessen replied:
I prefer the additional query :)
So you know what was deleted

nickvergessen commented:
This was also on my todo list, but thanks for picking it up!

ChristophWurst commented on tests/Settings/Controller/AuthSettingsControllerTest.php:
```diff
@@ -54,13 +55,16 @@ protected function setUp() {
 
-		$this->controller = new AuthSettingsController('core', $this->request, $this->tokenProvider, $this->userManager, $this->session, $this->secureRandom, $this->uid);
+		$activityManager = $this->createMock(IManager::class);
+		$activityManager->method('generateEvent')
```
It would be good if the calls on the activity manager are asserted where actually expected -> in the test relevant cases :wink:

danielkesselberg replied:
Have a look ;-)

ChristophWurst reviewed (approved):
Tests could be improved, but looks good :+1:

kesselb committed:
Add token name for delete activity

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

kesselb committed:
Add test for destroy with invalid user

Some code cleanup
- willReturn instead of will(returnValue)
- Annotation for mocked objects

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

danielkesselberg commented on settings/Controller/AuthSettingsController.php:
```diff
@@ -192,18 +216,55 @@ public function destroy($id) {
+	 * @throws \OC\Authentication\Exceptions\ExpiredTokenException
+	 */
+	private function findTokenByIdAndUser(int $id): IToken {
+		$token = $this->tokenProvider->getTokenById((string)$id);
```
@rullzer I'm a bit confused about `(string)$id`. Do you remember why you did it that way?

danielkesselberg replied:
Hmm. Looked at that again. There is no typehint for the route right no so basically $id could be anything but [getTokenById](https://github.com/nextcloud/server/blob/674930da7ff0a9bc82425caaf877efe6bae3bbde/lib/private/Authentication/Token/IProvider.php#L74) is typed. I guess this only works because `AuthSettingsController` is not strict yet so `23` is casted to `"23"` before passing to `getTokenById` where `"23"` is casted to `23` again.

danielkesselberg replied:
![image](https://user-images.githubusercontent.com/3902676/51985803-13142c80-249f-11e9-886f-d25fe6243d2f.png)

Does not work if `declare(strict_types=1)`. Let me remove `(string)` and hope that it does not :boom:

kesselb committed:
Don't cast tokenId to string

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

danielkesselberg commented:
Ready for another review ;-)

nickvergessen committed:
Use rich object subjects

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen committed:
Make strict

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen reviewed (approved):
Added rich subject and made the thing strict

danielkesselberg commented on settings/Activity/Provider.php:
```diff
@@ -141,15 +141,22 @@ protected function getParameters(IEvent $event) {
+			case self::APP_TOKEN_UPDATED:
+			case self::APP_TOKEN_DELETED:
+				return [
+					'token' => [
```
Cool :+1: I didn't understand how that works :see_no_evil: If there is no definition use [highlight](https://github.com/nextcloud/server/blob/422dd9dfd68bf678801a6ff1a1476d54d668df8f/lib/public/RichObjectStrings/Definitions.php#L259) :tada:

nickvergessen replied:
yeah, but only if it shouldnt have a definition. Tokens are nothing that has a page or something like that, so highlight is good enough.
